polly/lib/Analysis/ScopDetection.cpp

```diff
@@ -1621,6 +1626,12 @@ void ScopDetection::findScops(Region &R) {
     if (!ExpandedR)
       continue;
 
+    if (!isRegionExpansionProfitable(*ExpandedR, LI)) {
+      removeCachedResults(*ExpandedR);
+      delete ExpandedR;
```

Comment:
I think this should be moved inside `expandRegion` such that it returns nullptr when this test fails. Also avoids the manual memory management.

Reply:
Actually, region expansion profitability check should be done at the completion of `expandRegion`.

Because expanding a region to merge adjacent loops may involve multiple steps: 1) expand to include the exit block of current region, this added block may happen to be the preheader block of the next loop to be merged; 2) expand again to include all the blocks of the next loop; 3) repeat step1 and step2, or add other blocks from given CFG.

During the first expansion, the extra block added does not belong to any loops of the expandedRegion, but is actually the preheader block of the adjacent loop to be merged. Then the loop invariant loads for the next loop is going to cause the very first expansion to fail. Eventually region expansion will fail to merge with the adjacent loop, which should have been merged at the second expansion. 

test/CodeGen/region_expansion_profitability_check_loopfusion_1.ll added can be used to reproduce this issue.

Reply:
This passes all the tests (after fixing the 3 new ones):
```c++
Region *ScopDetection::expandRegion(Region &R) {
  ...
  while (ExpandedRegion) {
     ...
  }

  if (LastValidRegion && !isRegionExpansionProfitable(*LastValidRegion, LI)) {
    POLLY_DEBUG(dbgs() << "Non-profitable region: "
                       << LastValidRegion->getNameStr() << "\n");
    removeCachedResults(*LastValidRegion.get());
    return nullptr;
  }

  POLLY_DEBUG({
    if (LastValidRegion)
      dbgs() << "\tto " << LastValidRegion->getNameStr() << "\n";
    else
      dbgs() << "\tExpanding " << R.getNameStr() << " failed\n";
  });

  return LastValidRegion.release();
}
```